### ISSUE TITLE
InterfaceBuilder constraints conflict with programmatic constraints

### ISSUE THREAD

**Issue by acham1** [uchicago-mobi/mpcs51030-2017-summer-forum#19]
In the "Implement a Custom Control" section of **Start Developing iOS Apps**, we first drag a Horizontal Stack View from the Object Library to the StoryBoard, and then change the class to RatingControl via the Identity Inspector. This view has some arbitrary initial dimensions, which for me is 200x110.

Then we programmatically set the button dimensions in the RatingControl.swift file.

```
        button.translatesAutoresizingMaskIntoConstraints = false
        button.heightAnchor.constraint(equalToConstant: 44.0).isActive = true
        button.widthAnchor.constraint(equalToConstant: 44.0).isActive = true
        button.addTarget(self, action: #selector(RatingControl.ratingButtonTapped(button:)), for: .touchUpInside)
```

But when I build and run the project, I get the following error messages:
```

> 2017-07-07 06:31:50.019098-0500 FoodTracker[8547:380556] [LayoutConstraints] Unable to simultaneously satisfy constraints.
> 	Probably at least one of the constraints in the following list is one you don't want. 
> 	Try this: 
> 		(1) look at each constraint and try to figure out which you don't expect; 
> 		(2) find the code that added the unwanted constraint or constraints and fix it. 
> (
>     "<NSLayoutConstraint:0x608000095090 UIButton:0x7f80a9c0bf50.width == 44   (active)>",
>     "<NSLayoutConstraint:0x608000096e90 'UISV-canvas-connection' FoodTracker.RatingControl:0x7f80a9c0b320.leading == UIButton:0x7f80a9c0bf50.leading   (active)>",
>     "<NSLayoutConstraint:0x608000096f30 'UISV-canvas-connection' H:[UIButton:0x7f80a9c0bf50]-(0)-|   (active, names: '|':FoodTracker.RatingControl:0x7f80a9c0b320 )>",
>     "<NSLayoutConstraint:0x60800008f6e0 'UIView-Encapsulated-Layout-Width' FoodTracker.RatingControl:0x7f80a9c0b320.width == 200   (active)>"
> )
> 
> Will attempt to recover by breaking constraint 
> <NSLayoutConstraint:0x608000095090 UIButton:0x7f80a9c0bf50.width == 44   (active)>
> 
> Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
> The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
> 2017-07-07 06:31:50.020606-0500 FoodTracker[8547:380556] [LayoutConstraints] Unable to simultaneously satisfy constraints.
> 	Probably at least one of the constraints in the following list is one you don't want. 
> 	Try this: 
> 		(1) look at each constraint and try to figure out which you don't expect; 
> 		(2) find the code that added the unwanted constraint or constraints and fix it. 
> (
>     "<NSLayoutConstraint:0x608000095040 UIButton:0x7f80a9c0bf50.height == 44   (active)>",
>     "<NSLayoutConstraint:0x608000094fa0 'UISV-canvas-connection' FoodTracker.RatingControl:0x7f80a9c0b320.top == UIButton:0x7f80a9c0bf50.top   (active)>",
>     "<NSLayoutConstraint:0x608000097020 'UISV-canvas-connection' V:[UIButton:0x7f80a9c0bf50]-(0)-|   (active, names: '|':FoodTracker.RatingControl:0x7f80a9c0b320 )>",
>     "<NSLayoutConstraint:0x608000096ad0 'UIView-Encapsulated-Layout-Height' FoodTracker.RatingControl:0x7f80a9c0b320.height == 110   (active)>"
> )
> 
> Will attempt to recover by breaking constraint 
> <NSLayoutConstraint:0x608000095040 UIButton:0x7f80a9c0bf50.height == 44   (active)>
> 
> Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
> The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
> 
```
And my simulator ends up looking like this, showing that the programmatic constraints (44x44) are broken in favor of the arbitrary InterfaceBuilder/default ones.

<img src="https://user-images.githubusercontent.com/13604807/27956223-6961af10-62de-11e7-95ff-ea6f53fa29d3.png" width=200>

Is anyone else having this problem? And is there a way to cancel the 'UIView-Encapsulated-Layout-Width' constraints? The error says to "find the code that added the unwanted constraint", but I don't know where to do that, since I (following the tutorial) never added such code in the first place.

**Comment by prazgaitis** [comment]
I think its probably telling you that removing the following lines would fix the warning:

```button.heightAnchor.constraint(equalToConstant: 44.0).isActive = true```
```button.widthAnchor.constraint(equalToConstant: 44.0).isActive = true```

Alternatively, you could delete all the constraints in the storyboard and it should probably be fine. Is your project up to date? I'm happy to pull it down and take a look.

**Comment by tabinks** [comment]
It looks like you have some extra constraints set up.  IB doesn't like when there is a conflict, even it they are the same.  You can delete constraints by clicking on them and hitting delete or from the outline.  Click on a constraint and hit delete.  Just make sure you select the correct one.

FYI: Dealing with constraints is one of the least fun (and frustrating) part of developing for iOS.

**Comment by acham1** [comment]
@prazgaitis Thanks Paul, I do agree, but those are the actually the lines that I do want to keep, since they're directly from the tutorial instructions, and they keep the buttons small and square. Is there a way to quench the ` "<NSLayoutConstraint:0x60800008f6e0 'UIView-Encapsulated-Layout-Width' FoodTracker.RatingControl:0x7f80a9c0b320.width == 200   (active)>"` constraint instead? Also, yes please, my repo is up-to-date; could you please pull it down and take a look?

<img src="https://user-images.githubusercontent.com/13604807/27980749-b18155c0-6347-11e7-934f-70ad84ff981a.png" width=200>

@tabinks Thanks Andrew, I agree as well, but in this case the problematic IB constraints in question aren't listed under the blue/constraints/drop-down icon in project navigator, so I'm not sure how to go about deleting them? I think the constraints are from the Size Inspector fields for height and width, but if I delete the entries, it just changes the constraint to size 0.

**Comment by hbennett766** [comment]
Is your repo up to date? I can pull down your code and take a look.

**Comment by acham1** [comment]
Yea, it's up to date; thanks!

**Comment by acham1** [comment]
I can kind of work around it by manually setting the horizontal stack view dimensions to 252x44 in Size Inspector, and then the warnings/errors go away. I just want to double-check that there isn't a better way? I had originally thought that the custom view could dynamically and programmatically change its dimensions, irrespective of the fixed values in Size Inspector?

**Comment by acham1** [comment]
Sorry for the trouble! It seems the problem is resolved now; I think I just had to first explicitly add some auto-layout constraints for horizontal/vertical positions (these steps weren't mentioned in the tutorial), and then the horizontal stack view dimensions began to update automatically the way we expect it to. 

So, oddly enough, I think the issue was that I had set too few (position) constraints, rather than too many (size) constraints. Somehow, the layout was treating the drawing/icon of the StackView in InterfaceBuilder as a fixed dimension constraints. I didn't see any of the usual warnings about ambiguous positions, maybe because I had lined up the view using the purple/blue guide-lines that show up when dragging. Anyhow, adding explicit position constraints seems to have fixed it.

yay :-)
